### ISSUE TITLE
Execute delayed configuration closures only once.

### ISSUE THREAD

**Issue by mkmaier** [gradle/gradle#4292]
### Expected Behavior
Closures used to delay configuration of task properties should only be executed once, not multiple times.

### Current Behavior
Depending of the task and property that is configured, the closure is executed up to four (4) times (for `Copy.from` when not upto-date).

### Context
Delayed configuration is often used to avoid expensive operations during configuration phase, so it should not cause those operations to be executed multiple times.

### Steps to Reproduce (for bugs)
Unzip [gradle_bug.zip](https://github.com/gradle/gradle/files/1703570/gradle_bug.zip), run `gradle wrapper`, `gradlew copy`, `gradlew clean` and see duplicated output.

**Comment by wolfs** [comment]
Hi @mkmaier, thank you very much for your report.

as a workaround, you could move your expensive calculation into a so called configure task, which configures the copy task. You may be able to use memoization to store the result of your expensive calculation.

If you try out your example project with Gradle 4.4, then you see that `into closure` is printed 10 times, as opposed to the 3 times with Gradle 4.5.1. We actually improved in this area.

While I understand that you expect your configuration closure to only be executed once, Gradle needs to evaluate the inputs at different times:
- when determining the dependencies of the task graph
- just before executing the task to determine the actual inputs
- the task itself evaluates the configuration when it executes, too.

We cannot just assume that the property does not change between building the task graph and actually running the task mostly for backwards compatibility and since some use cases require changing the configuration at runtime (like configure tasks).

We also don't adjust the task implementation in a way that it would use the evaluated configuration we tracked for incremental build.

So it seems like three times would be the minimum for now.

To sum up: Gradle could do better, but it is not trivial. If you have some ideas how to make this better, please provide comments on this issue or a pull request.

**Comment by mkmaier** [comment]
Thank you for your quick and comprehensive response, @wolfs!

I do not see any noticeable performance problems, since we mostly have resolves and maybe a zipTree in our config closures, both of which should be cacheable, but I assumed (wrongly, as it turns out) that the closures would be "performance critical" (and thus should only be executed once).
Looking for documentation, I could only find a short post by Mr. Haki and several forum mentions, but no authoritative explanation. Maybe the user documentation should be extended with an explanation of the various lazy configuration techniques, or it should be made easier to find if it is there.

I'd be willing to write something up for you to extend/correct if necessary, if you could point me at the right place to do this.

**Comment by wolfs** [comment]
@mkmaier Thank you for your offer to write something up! https://docs.gradle.org/current/userguide/lazy_configuration.html seems like a good URL, the fact about the number of times a lazy property is evaluated applies to `Provider`s, too.
You can either provide a PR or write something up and post it as a comment to this issue and then we can find a good place in the user guide for the content.

**Comment by grv87** [comment]
According to my test in #5301, property closure was called once in Gradle 4.2.1, and that was changed in 4.3 RC 1.
So, it was a bug in 4.2.1, and now everything works correctly?

**Comment by wolfs** [comment]
@grv87 Looks like we need to call the `Closure` for a property registered via the runtime API once more for validation in `ValidatingTaskExecuter`. That is unnecessary. We may fix this in a future Gradle release.
